tests/v2/test_waku_lightpush.nim

```diff
@@ -5,78 +5,119 @@ import
-    discard
+  await proto.start()
+  switch.mount(proto)
 
-  asyncTest "handle light push request fail":
-    let
-      key = PrivateKey.random(ECDSA, rng[]).get()
-      listenSwitch = newStandardSwitch(some(key))
-    await listenSwitch.start()
+  return proto
 
-    let dialSwitch = newStandardSwitch()
-    await dialSwitch.start()
+proc newTestWakuLightpushClient(switch: Switch): WakuLightPushClient =
+  let
+    peerManager = PeerManager.new(switch)
+    rng = crypto.newRng()
+  WakuLightPushClient.new(peerManager, rng)
 
 
-    proc requestHandler(requestId: string, msg: PushRequest) {.gcsafe, closure.} =
-      # TODO Success return here
-      debug "handle push req"
-      check:
-        1 == 0
+suite "Waku Lightpush":
 
-    # FIXME Unclear how we want to use subscriptions, if at all
-    let
-      peerManager = PeerManager.new(dialSwitch)
-      rng = crypto.newRng()
-      proto = WakuLightPush.init(peerManager, rng, requestHandler)
+  asyncTest "push message to pubsub topic is successful":
+    ## Setup
+    let 
+      serverSwitch = newTestSwitch()
+      clientSwitch = newTestSwitch()
+
+    await allFutures(serverSwitch.start(), clientSwitch.start())
 
-    proto.setPeer(listenSwitch.peerInfo.toRemotePeerInfo())
-    waitFor proto.start()
-    dialSwitch.mount(proto)
+    ## Given
+    let handlerFuture = newFuture[(string, WakuMessage)]()
+    let handler = proc(peer: PeerId, pubsubTopic: string, message: WakuMessage): Future[WakuLightPushResult[void]] {.async.} = 
+        handlerFuture.complete((pubsubTopic, message))
+        return ok()
 
+    let
+      server = await newTestWakuLightpushNode(serverSwitch, handler)
+      client = newTestWakuLightpushClient(clientSwitch)
 
-    # TODO Can possibly get rid of this if it isn't dynamic
-    proc requestHandler2(requestId: string, msg: PushRequest) {.gcsafe, closure.} =
-      debug "push request handler"
-      # TODO: Also relay message
-      # TODO: Here we want to send back response with is_success true
-      discard
+    let serverPeerId = serverSwitch.peerInfo.toRemotePeerInfo()
 
     let
-      peerManager2 = PeerManager.new(listenSwitch)
-      rng2 = crypto.newRng() 
-      proto2 = WakuLightPush.init(peerManager2, rng2, requestHandler2)
-    waitFor proto2.start()
-    listenSwitch.mount(proto2)
+      topic = DefaultPubsubTopic
+      message = fakeWakuMessage()
+
+    ## When
+    let rpc = PushRequest(pubSubTopic: topic, message: message)
+    let requestRes = await client.request(rpc, serverPeerId)
+
+    require await handlerFuture.withTimeout(100.millis)
+
+    ## Then
+    check:
+      requestRes.isOk()
+      handlerFuture.finished()
 
+    let (handledMessagePubsubTopic, handledMessage) = handlerFuture.read()
+    check:
+      handledMessagePubsubTopic == topic
+      handledMessage == message
+
+    ## Cleanup
+    await allFutures(clientSwitch.stop(), serverSwitch.stop())
```

Comment:
Agree. We have not explored the right way of implementing test case setup and cleanup (for example, Nim unittest library supports setup and teardown procedures).

In any case, the way that `require` works, IIRC, it terminates the whole test execution if the condition is not `true`. So, the GC will teardown any resources dangling.

For consistency, I would like to keep it like it is RN. And change it in all the test suites once we have a clean way of implementing it.

Reply:
P.S. I come from Golang world, I'd rather not to use `defer` unless it is strictly necessary

Reply:
sure. i was concerned since beyond this particular example, we can have a case where the resources are not automatically freed/cleaned. i.e. a file being created that is not removed. nothing critical though.

setup and teardown might be overkill in some cases, but happy to start the discussions around it.

resolving :)